### DISCUSSION TITLE
is there a way to get hover on demand?

### DISCUSSION THREAD

**Discussion by fommil** [joaotavora/eglot#1328] [General]
I have previously enjoyed the pre-LSP workflow in ensime (scala), haskell (hsinspect) and golang (`godef-describe`) of having a key binding to echo the type information at point to the minibuffer.

Enabling the server feature `:hover` gives roughly the same feature (perhaps gives too much information at times) but unfortunately is making a lot of requests because it is communicating with the server every time I move the point.

I have also looked into the `eldoc-doc-buffer` command but it opens the information in a buffer and not the minibuffer. It also doesn't work so well if `:hoverProvider` is disabled and will sometimes give me the information I wanted but the rest of the time opens up an empty `*eldoc*` buffer or a stale one from whatever previous eldoc I had asked for (including in other modes, such as if I am editing a `.el` file). This also happens for `eldoc-print-current-symbol-info`.

Is it possible to disable automatic hover events being sent to the server, and insted only trigger them manually from an interactive function that I can bind to a key combination?

**Comment by joaotavora** [comment]
`M-x eldoc-mode` and then `M-x eldoc` manually?  This isn't an Eglot issue, btw

**Reply by fommil** [reply]
@joaotavora 🤦🏻 I didn't realise that `eldoc` was creating the `hover` events! Ok, that works great, thank you!

I think I am almost in a position to delete all my custom emacs code for the various LSP servers I've written over the years, and move to using eglot for everything. Thank you for your work on this! When I started out all the LSP clients were so heavyweight I ran away and wrote direct code invocation, but I think eglot hits a really good sweet spot.

**Reply by fommil** [reply]
Figuring out when to turn off eldoc in the buffer startup sequence is going to be interesting... also, I know this is `eldoc` config, but I can't figure out how to get it to use the minibuffer instead of a popup buffer.

**Reply by fommil** [reply]
Strangely, when using `hover`, the docs are always sent to the minibuffer, but when using `eldoc` they always go to the buffer. And if `eldoc` is called on something that is already in the minibuffer, it appears in both!

**Reply by fommil** [reply]
Related: https://github.com/casouri/eldoc-box

Although it uses a popup box and I'd really rather it used the minibuffer, but I can get used to it. It also doesn't seem to be able to disable the hovering behaviour.

**Reply by fommil** [reply]
Somebody else with the same question, but no answer, https://emacs.stackexchange.com/questions/73983/how-to-make-eldoc-only-popup-on-demand

**Reply by fommil** [reply]
This seems to work for getting on-demand "hover" info, as you suggested:

```
(add-hook
 'eglot-managed-mode-hook
 (lambda ()
   ;; we want eglot to setup callbacks from eldoc, but we don't want eldoc
   ;; running after every command. As a workaround, we disable it after we just
   ;; enabled it. Now calling `M-x eldoc` will put the help we want in the eldoc
   ;; buffer. Alternatively we could tell eglot to stay out of eldoc, and add the
   ;; hooks manually, but that seems fragile to updates in eglot.
   (eldoc-mode -1)))
```

But I can't figure out how to get `eldoc` to prefer to use the minibuffer. By adding it to `eldoc-add-command` I end up getting the info in the minibuffer AND the eldoc buffer.

**Reply by joaotavora** [reply]
This might need a patch to Emacs's `eldoc.el` library:
```diff
diff --git a/lisp/emacs-lisp/eldoc.el b/lisp/emacs-lisp/eldoc.el
index 22144ed7c18..36b0eb52511 100644
--- a/lisp/emacs-lisp/eldoc.el
+++ b/lisp/emacs-lisp/eldoc.el
@@ -605,7 +605,7 @@ eldoc--echo-area-prefer-doc-buffer-p
                     'maybe)))
        (get-buffer-window eldoc--doc-buffer t)))
 
-(defun eldoc-display-in-echo-area (docs _interactive)
+(defun eldoc-display-in-echo-area (docs interactive)
   "Display DOCS in echo area.
 Honor `eldoc-echo-area-use-multiline-p' and
 `eldoc-echo-area-prefer-doc-buffer'."
@@ -614,11 +614,13 @@ eldoc-display-in-echo-area
     ;; example, if this-command is non-nil while running via an idle
     ;; timer, we're still in the middle of executing a command, e.g. a
     ;; query-replace where it would be annoying to overwrite the echo
-    ;; area.
-    (or
-     (not (eldoc-display-message-no-interference-p))
-     this-command
-     (not (eldoc--message-command-p last-command))))
+    ;; area.  Interactively (i.e. via M-x eldoc), we don't care about
+    ;; this.
+    (and (not interactive)
+         (or
+          (not (eldoc-display-message-no-interference-p))
+          this-command
+          (not (eldoc--message-command-p last-command)))))
    (;; If we do but nothing to report, clear the echo area.
     (null docs)
     (eldoc--message nil))
```

Then you need to read up on `eldoc-display-functions`, and I think you want to keep only  the echo-area one (what you call "minibuffer") in there.

**Reply by fommil** [reply]
@joaotavora thanks for looking into this. It is unfortunate for sure. I guess my options are to monkey patch the function locally or just get used to the behaviour. I'll try `eldoc-box`, and maybe I'll take a look at how they hijack the display... maybe I can add something into `eldoc-display-functions` if it really bothers me eventually.

**Reply by joaotavora** [reply]
>  It is unfortunate for sure. 

No, not unfortunate, I've just pushed that fix to Emacs's `eldoc.el` !  So now you just need to upgrade to ElDoc 1.15 and setup `eldoc-display-functions` like I suggested and then `M-x eldoc` will work like you want.

**Reply by fommil** [reply]
@joaotavora oh wow, thank you! I didn't think a change like that would stand a chance of getting included in core!

**Reply by fommil** [reply]
is this not going to annoy people who want to run `M-x eldoc` and have it pop up?

**Reply by fommil** [reply]
This works well, I might write a wrapper function and local binding for `eldoc-display-functions` that decides sets whether to use the echo area or the buffer depending on whether I provide a prefix argument or not. So I get echo area by default (which tendns to have truncated information) but I get the full info if I pass the C-u prefix.

**Reply by fommil** [reply]
@joaotavora this works nicely, I've probably butchered lexical vs dynamic scoping and the design architecture of eldoc to get this to work, but ... here it is....

```
(defun eldoc-fancy (arg)
  "`eldoc' but uses the echo area by default and a prefix will swap to a buffer."
  (interactive "P")
  (let (_)
    (defvar eldoc-display-functions)
    (setq eldoc-display-functions
          (if arg '(eldoc-display-in-buffer) '(eldoc-display-in-echo-area)))
    (call-interactively #'eldoc)))
```

**Reply by joaotavora** [reply]
```
(defun eldoc-fancy (arg)
  "`eldoc' but uses the echo area by default and a prefix will swap to a buffer."
  (interactive "P")
  (let ((eldoc-display-functions
          (if arg '(eldoc-display-in-buffer) '(eldoc-display-in-echo-area))))
    (eldoc t)))
```

**Reply by fommil** [reply]
@joaotavora that `let` binding is what I originally wrote but I have lexical-binding by default so I wanted to force dynamic here. Thanks for the `(eldoc t)` tip here.

**Reply by joaotavora** [reply]
It doesn't matter, it will work.  Use `let`, don't "force dynamic"

**Reply by fommil** [reply]
@joaotavora hmm, it doesn't seem to rebind the variable when I do it your way and so I get it in the echo area and in a buffer, regardless of how I invoke the command. I noticed this when I first wrote it but I thought I'd try again since you suggested. The file where I define this has

```
;;; init.el --- Emacs configuration -*- lexical-binding: t; -*-
````

and of course so does

```
;;; eldoc.el --- Show function arglist or variable docstring in echo area  -*- lexical-binding:t; -*-
```

This is in a realm where I don't don't feel confident relying on my understanding of the theory, and I rely on experiment to guide me.

**Reply by joaotavora** [reply]
makes absolutely no difference here, just tested, but suit yourself

**Reply by fommil** [reply]
I must have something defined in my config that is interfering here 🤔, I have no idea why one form works and the other does not. My understanding of the theory is that they should be the same, but yet it only works for me when I force dynamic. The variable is global for me, perhaps it needs to be make a local variable?

**Reply by joaotavora** [reply]
```
$ cat ~/tmp/foo.el
;;; foo.el --- asd                                   -*- lexical-binding: t; -*-
(global-eldoc-mode -1)

(defun eldoc-fancy (arg)
  "`eldoc' but uses the echo area by default and a prefix will swap to a buffer."
  (interactive "P")
  (let ((eldoc-display-functions
          (if arg '(eldoc-display-in-buffer) '(eldoc-display-in-echo-area))))
    (eldoc t)))


$ src/emacs -Q -l ~/tmp/foo.el

type (defun
type M-x eldoc-fancy
type C-u M-x eldoc-fancy
```

**Reply by fommil** [reply]
that works, so definitely something weird in my personal setup.
